### PR TITLE
Fix commit hooks without staged files

### PR DESCRIPTION
**Pull request by lstrojny**
Changes `HooksResolver.filterWithVCS()` so that it only filters for hooks that will have staging areas.

Fixes https://github.com/Escape-Technologies/mookme/issues/63

### CODE REVIEW

File: packages/mookme/src/loaders/hooks-resolver.ts [hunk 1]
```diff
@@ -18,9 +18,10 @@ function matchExactPath(filePath: string, to_match: string): boolean {
  * A class defining several utilitaries used to load and prepare packages hooks to be executed
  */
 export class HooksResolver {
-  gitToolkit: GitToolkit;
-  root: string;
-  hookType: string;
+  private static readonly stagingHooks = ['pre-commit'];
+  private readonly gitToolkit: GitToolkit;
+  private readonly root: string;
+  private readonly hookType: string;
```
Comment (LMaxence) [inline packages/mookme/src/loaders/hooks-resolver.ts:24]: Is it necessary to turn this into privates ? It is very java-ish and these property are used elsewhere (the `hookType` property for instance) ?

I'm not especially against this syntax (though I don't think private + getter/setters are necessary here) but I am a much bigger fan of consistency

File: packages/mookme/src/utils/git.ts [hunk 2]
```diff
@@ -28,17 +28,18 @@ export class GitToolkit {
...
+  }
+
+  private static gitFileList(command: string): string[] {
+    return execSync(command)
       .toString()
-      .split(' ')
-      .map((pth) => pth.replace('\n', ''));
+      .split('\0')
+      .filter((p) => p !== '');
   }
```
Comment (LMaxence) [inline packages/mookme/src/utils/git.ts:43]: 👍

File: packages/mookme/tests/cases/post-commit-succeeds.sh [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+git add .
+git commit -m "Set up hooks"
+
+node $ROOT_FOLDER/dist/index.js init \
+    --yes \
+    --added-behaviour exit \
+    --skip-types-selection
+
+node $ROOT_FOLDER/dist/index.js run -t post-commit
```
Comment (LMaxence) [inline packages/mookme/tests/cases/post-commit-succeeds.sh:35]: Thanks for adding a test !

**Comment by lstrojny** [comment]
Thinking about this more, the implementation is wrong. For post-commit, filtering needs to happen, it’s just that it needs to happen on the previously committed changeset, so changed files in `HEAD^`. What is needed is different strategies to detect changes based on what hook is invoked.

**Comment by lstrojny** [comment]
While digging deeper, I found another issue with partial commits. Say you use `git add -p` to partially stage a file, that will not be handled properly by the linter. [lint-staged](https://github.com/okonet/lint-staged) does that beautifully, what do you think about integrating lint-staged as a library for the `pre-commit` hook?

**Comment by LMaxence** [comment]
> Thinking about this more, the implementation is wrong. For post-commit, filtering needs to happen, it’s just that it needs to happen on the previously committed changeset, so changed files in `HEAD^`. What is needed is different strategies to detect changes based on what hook is invoked.

This is a very good idea. I'll implement a system for having multiple strategies for selecting files for the comparison next week, this refacto is bigger than expected.

One can even imagine a step option where the user gets to decide which strategy to use for a specific step, even though thing are done the other way around so far (the default strategy, comparison against the staged files, is done before loading the concerned steps)

**Comment by LMaxence** [comment]
I quickly came up with #66, solving the urge for a rework of the way we choose how to filter the hooks to run. I added a basis for a strategy `PreviousCommitFilterStrategy`, where you only have to incorporate the logic for actually selecting the hooks to run.

Feel free to rely on the example provided in `CurrentCommitFilterStrategy` (the one used by default for every hook type so far)

**Comment by LMaxence** [comment]
I'm closing this as a fix candidate has been merged in beta :)

**Pull request closed without merging**